src/core/thread/network_diagnostic.cpp

```diff
@@ -707,6 +707,290 @@ void NetworkDiagnostic::HandleDiagnosticReset(Coap::Message &aMessage, const Ip6
+        aNetworkDiagRoute.mRouteData[routeCount].mLinkQualityOut = aRouteTlv.GetLinkQualityOut(routeCount);
+        ++routeCount;
+    }
+    aNetworkDiagRoute.mRouteCount = routeCount;
+}
+
+static inline void ParseLeaderData(const LeaderDataTlv &aLeaderDataTlv, otLeaderData &aLeaderData)
+{
+    aLeaderData.mPartitionId       = aLeaderDataTlv.GetPartitionId();
+    aLeaderData.mWeighting         = aLeaderDataTlv.GetWeighting();
+    aLeaderData.mDataVersion       = aLeaderDataTlv.GetDataVersion();
+    aLeaderData.mStableDataVersion = aLeaderDataTlv.GetStableDataVersion();
+    aLeaderData.mLeaderRouterId    = aLeaderDataTlv.GetLeaderRouterId();
+}
+
+static inline void ParseMacCounters(const MacCountersTlv &aMacCountersTlv, otNetworkDiagMacCounters &aMacCounters)
+{
+    aMacCounters.mIfInUnknownProtos  = aMacCountersTlv.GetIfInUnknownProtos();
+    aMacCounters.mIfInErrors         = aMacCountersTlv.GetIfInErrors();
+    aMacCounters.mIfOutErrors        = aMacCountersTlv.GetIfOutErrors();
+    aMacCounters.mIfInUcastPkts      = aMacCountersTlv.GetIfInUcastPkts();
+    aMacCounters.mIfInBroadcastPkts  = aMacCountersTlv.GetIfInBroadcastPkts();
+    aMacCounters.mIfInDiscards       = aMacCountersTlv.GetIfInDiscards();
+    aMacCounters.mIfOutUcastPkts     = aMacCountersTlv.GetIfOutUcastPkts();
+    aMacCounters.mIfOutBroadcastPkts = aMacCountersTlv.GetIfOutBroadcastPkts();
+    aMacCounters.mIfOutDiscards      = aMacCountersTlv.GetIfOutDiscards();
+}
+
+static inline void ParseChildEntry(const ChildTableEntry &aChildTableTlvEntry, otNetworkDiagChildEntry &aChildEntry)
+{
+    aChildEntry.mTimeout = aChildTableTlvEntry.GetTimeout();
+    aChildEntry.mChildId = aChildTableTlvEntry.GetChildId();
+    ParseMode(aChildTableTlvEntry.GetMode(), aChildEntry.mMode);
+}
+
+otError NetworkDiagnostic::GetNextDiagTlv(const otMessage &      aMessage,
+                                          otNetworkDiagIterator &aIterator,
+                                          otNetworkDiagTlv &     aNetworkDiagTlv)
+{
+    otError              error   = OT_ERROR_PARSE;
+    const Coap::Message &message = static_cast<const Coap::Message &>(aMessage);
+    uint16_t             offset  = message.GetOffset();
+    uint16_t             tlvTotalLength;
+    NetworkDiagnosticTlv tlv;
+
+    VerifyOrExit(aIterator + offset < message.GetLength(), error = OT_ERROR_NOT_FOUND);
+    offset += aIterator;
+
+    VerifyOrExit(message.Read(offset, sizeof(tlv), &tlv) == sizeof(tlv));
+    VerifyOrExit(!tlv.IsExtended());
+
+    switch (tlv.GetType())
+    {
+    case NetworkDiagnosticTlv::kExtMacAddress:
+    {
+        ExtMacAddressTlv extMacAddr;
+
+        tlvTotalLength = sizeof(extMacAddr);
+        VerifyOrExit(message.Read(offset, tlvTotalLength, &extMacAddr) == tlvTotalLength);
+        VerifyOrExit(extMacAddr.IsValid());
+        ParseExtMacAddr(extMacAddr, aNetworkDiagTlv.mExtAddress);
```

Comment:
I think we can use this instead ?
```suggestion
        aNetworkDiagTlv.mExtAddress = *extMacAddr.GetMacAddr();
```
(not sure why `ExtMacAddressTlv ::GetMacAddr()` is returning a pointer instead of a const reference).